substrate/frame/support/procedural/src/pallet/parse/call.rs

```diff
@@ -269,57 +290,47 @@ impl CallDef {
 				}
 
 				let cfg_attrs: Vec<syn::Attribute> = helper::get_item_cfg_attrs(&method.attrs);
-				let mut call_idx_attrs = vec![];
-				let mut weight_attrs = vec![];
-				let mut feeless_attrs = vec![];
+				let mut call_index = None;
+				let mut weight = None;
+				let mut feeless_check = None;
+
 				for attr in helper::take_item_pallet_attrs(&mut method.attrs)?.into_iter() {
 					match attr {
-						FunctionAttr::CallIndex(_) => {
-							call_idx_attrs.push(attr);
+						FunctionAttr::CallIndex(idx) => {
+							if call_index.is_some() {
+								let msg =
+									"Invalid pallet::call, too many call_index attributes given";
+								return Err(syn::Error::new(method.sig.span(), msg))
+							}
+
+							call_index = Some(idx);
 						},
-						FunctionAttr::Weight(_) => {
-							weight_attrs.push(attr);
+						FunctionAttr::Weight(w) => {
+							if weight.is_some() {
+								let msg = "Invalid pallet::call, too many weight attributes given";
+								return Err(syn::Error::new(method.sig.span(), msg))
+							}
+							weight = Some(w);
 						},
-						FunctionAttr::FeelessIf(span, _) => {
-							feeless_attrs.push((span, attr));
+						FunctionAttr::FeelessIf(span, closure) => {
+							if feeless_check.is_some() {
+								let msg =
+									"Invalid pallet::call, there can only be one feeless_if attribute";
+								return Err(syn::Error::new(span, msg))
+							}
+
+							feeless_check = Some(closure);
 						},
 					}
 				}
 
-				if weight_attrs.is_empty() && dev_mode {
-					// inject a default O(1) weight when dev mode is enabled and no weight has
-					// been specified on the call
-					let empty_weight: syn::Expr = syn::parse_quote!(0);
-					weight_attrs.push(FunctionAttr::Weight(empty_weight));
-				}
```

Comment:
This logic is removed. It is redundant with the match below, which is refactored into `CallWeightDef::try_from`